webapp/src/js/services/rules-engine.js

```diff
@@ -1,256 +1,182 @@
-        _.values(callbacks).forEach(function(cbs) {
-          _.values(cbs).forEach(function(callback) {
-            callback(err);
-          });
-        });
-      };
-
-      var notifyCallbacks = function(fact, type) {
-        if (!emissions[type]) {
-          emissions[type] = {};
-        }
-        emissions[type][fact._id] = fact;
-        _.values(callbacks[type]).forEach(function(callback) {
-          callback(null, [ fact ]);
-        });
-      };
-
-      var assertFacts = function() {
-        KNOWN_TYPES.forEach(function(type) {
-          session.on(type, function(fact) {
-            notifyCallbacks(fact, type);
-          });
-        });
-        facts.forEach(function(fact) {
-          session.assert(fact);
-        });
-        session.matchUntilHalt().then(
-          // halt
-          function() {
-            notifyError(new Error('Unexpected halt in fact assertion.'));
-          },
-          // error
-          notifyError
-        );
-      };
+      const contactsAreIds = typeof contacts[0] === 'string';
+      const contactIds = contacts.map(contact => contact._id || contact);
+      const fetchContactDocs = contactIds => DB().allDocs({ keys: contactIds, include_docs: true })
+        .then(result => result.rows.map(row => row.doc).filter(existing => existing));
+      const fetchedContactDocs = contactsAreIds ? fetchContactDocs(contacts) : $q.resolve(contacts);
 
-      var findFact = function(id) {
-        return _.find(facts, function(fact) {
-          return contactHasId(fact.contact, id) ||
-                 _.findWhere(fact.reports, { _id: id });
+      return fetchedContactDocs
+        .then(contactDocs => {
+          const subjectIds = contactDocs.reduce((agg, contactDoc) => [...agg, ...registrationUtils.getSubjectIds(contactDoc)], []);
+          return $q.all([
+            Search('reports', { subjectIds }, { include_docs: true }),
+            fetchTasksFor(contactIds, 'requester'),
+          ])
+            .then(([reportDocs, taskDocs]) => ({
+              contactDocs,
+              reportDocs,
+              taskDocs,
+            }));
         });
-      };
+    },
+    
+    /*
+    Why two functions to fetch tasks?
+    PouchDB.query slows down when provided with a large keys array. 
+    For users with ~1000 users it is ~50x faster to provider a single start/end key.
+    */
+    allTasks: () => DB().query('medic-client/tasks', { startkey: 'owner-', endkey: 'owner-\ufff0', include_docs: true })
+      .then(res => res.rows.map(row => row.doc)),
+    
+    tasksOwnedBy: contactIds => fetchTasksFor(contactIds, 'owner'),
+
+    allContactDocs: function () {
+      return ContactTypes.getPersonTypes()
+      .then(function(personTypes) {
+        const keys = personTypes.map(personType => ([personType.id]));
+        return DB()
+          .query('medic-client/doc_by_type', { keys, include_docs: true })
+          .then(res => res.rows.map(row => row.doc).filter(doc => doc));
```

Comment:
This won't work for configurable hierarchy as all contacts will have doc.type = `contact`. Consider using `contacts_by_type` instead.